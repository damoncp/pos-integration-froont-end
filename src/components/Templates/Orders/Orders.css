#orders {
  border-left: 1px solid #ddd;
  height: 100%;
}

.orders-text__regular {
  font-size: .8rem;
}

.orders-table__row {
  vertical-align: middle;
}

.orders-table__row td:first-child {
  display: flex;
  align-items: center;
}

/* overwrite default css */
#orders .container {
  height: 100%;
  display: flex;
  flex-direction: column;
}
#orders .container > .row:first-child {
  flex-grow: 1;
}

#orders .table > :not(caption) > * > * {
  border-bottom-width: 0px;
}

#orders .table > :not(:last-child) > :last-child > * {
  border-bottom-color: #ddd;
}

#orders button,
#orders button:disabled {
  background-color: transparent;
  text-decoration: none;
  color: black;
  border: none;
  outline: none;
  box-shadow: none;
  padding: 0 .5rem;
}

#orders button:hover,
#orders button:focus,
#orders button:active {
  background-color: transparent;
  color: black;
  border: none;
  outline: none;
  box-shadow: none;
}
