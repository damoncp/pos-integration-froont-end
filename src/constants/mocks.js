const mockProducts = [
  {
    type: "burger",
    data: [
      {
        title: 'Chicken Burger',
        image: 'https://media.istockphoto.com/photos/bacon-burger-picture-id520215281?s=612x612',
        price: 12,
      },
      {
        title: 'Vegan Burger',
        image: 'https://media.istockphoto.com/photos/bacon-burger-picture-id520215281?s=612x612',
        price: 12,
      },
      {
        title: 'Beef Burger',
        image: 'https://media.istockphoto.com/photos/bacon-burger-picture-id520215281?s=612x612',
        price: 12,
      },
      {
        title: 'Lamb Burger',
        image: 'https://media.istockphoto.com/photos/bacon-burger-picture-id520215281?s=612x612',
        price: 12,
      },
    ]
  },
  {
    type: "fries",
    data: [
      {
        title: 'Small Fries',
        image: 'https://media.istockphoto.com/photos/french-fries-with-sauce-picture-id1130883810',
        price: 7,
      },
      {
        title: 'Large Fries',
        image: 'https://media.istockphoto.com/photos/french-fries-with-sauce-picture-id1130883810',
        price: 7,
      },
    ]
  },
  {
    type: "drinks",
    data: [
      {
        title: 'Coca Cola',
        image: 'https://cdn.pixabay.com/photo/2015/01/08/04/18/box-592367_1280.jpg',
        price: 3,
      },
    ]
  },
];

export {
  mockProducts,
};
