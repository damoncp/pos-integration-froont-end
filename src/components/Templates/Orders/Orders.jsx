import './Orders.css';
import PropTypes from 'prop-types';
import React, { useContext, useEffect, useState } from 'react';
import { Button, Col, Container, Row, Spinner, Table } from 'reactstrap';
import GlobalContexts from '../../../contexts/globalContexts';
// utils
import {
  decrement,
  increment,
  getProductsList,
  payButtonDisabled,
} from '../../../utils/ordersHelpers';
import payRequest from '../../../utils/payRequest';

export default function Orders() {
  const [currentProducts, setCurrentProducts] = useState([]);
  const [totalCost, setTotalCost] = useState(0);
  const [loading, setLoading] = useState(false);
  const [error, setError] = useState('');

  const { contexts: { products } } = useContext(GlobalContexts);

  useEffect(() => {
    // convert products object array into products array
    const currentProductsList = getProductsList(products);
    setCurrentProducts(currentProductsList);
  }, [products])

  const updateQuantityValue = (operation, index) => {
    let updatedProductsList = [...currentProducts];
    let targetProduct = currentProducts[index];

    targetProduct.quantity = operation(targetProduct.quantity);
    updatedProductsList[index] = targetProduct;

    setCurrentProducts(updatedProductsList);
  }

  const updateTotalCost = () => {
    // calculate total payment first
    const eachItemCost = currentProducts
      .map(({ name, price, quantity }) => price * quantity);

    setTotalCost(eachItemCost.reduce(
      (accumulator, value) => accumulator + value
    ));
  }

  const quantityDecrease = (index) => {
    updateQuantityValue(decrement, index);
    updateTotalCost();
  };

  const quantityIncrease = (index) => {
    updateQuantityValue(increment, index);
    updateTotalCost();
  };

  return (
    <div id="orders">
      <h5 className={"text--centered text--bold heading__default"}>
        Order
      </h5>
      <Container>
        <Row>
          <Col>
            <Table responsive>
              <thead>
                <tr>
                  <th>Quantity</th>
                  <th>Name</th>
                  <th>Price</th>
                </tr>
              </thead>
              <tbody>
                {currentProducts.length === 0 ?
                  (<tr>
                      <td>
                        <Spinner color={"dark"} children="" />
                      </td>
                    </tr>)
                  : currentProducts.map(({ title, quantity, price }, index) => (
                    <tr key={`${index}--${title}`} className="orders-table__row">
                      <td>
                        <Button
                          onClick={() => { quantityDecrease(index) }}
                          disabled={quantity === 0}
                        >-</Button>
                        <span className="orders-text__regular">{quantity}</span>
                        <Button
                          onClick={() => { quantityIncrease(index) }}
                        >+</Button>
                      </td>
                      <td className="orders-text__regular">{title}</td>
                      <td className="orders-text__regular">${price}</td>
                    </tr>
                  ))}
              </tbody>
            </Table>
          </Col>
        </Row>
        {error && (
          <Row>
            <Col>{error}</Col>
          </Row>)}
        <Row>
          <Col>Total: </Col>
          <Col data-testid="order-total">$ {totalCost}</Col>
        </Row>
        <Row>
          <Button
            className="element--no-border-radius bg-success text-white p-3 mt-2"
            onClick={() => payRequest(
              setLoading,
              setError,
              setTotalCost,
              setCurrentProducts,
              totalCost,
              currentProducts,
              products)}
            disabled={payButtonDisabled(loading, totalCost)}
          >
            {loading ? <Spinner color={"dark"} children="" /> : 'Pay'}
          </Button>
        </Row>
      </Container>
    </div>
  )
}

Orders.propTypes = {
  contexts: PropTypes.shape({
    products: PropTypes.arrayOf({
      type: PropTypes.string,
      data: PropTypes.arrayOf({
        title: PropTypes.string,
        image: PropTypes.string,
        price: PropTypes.number,
      })
    })
  }),
};
