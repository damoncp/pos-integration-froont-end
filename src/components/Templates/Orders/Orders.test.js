import { render, screen, cleanup,fireEvent } from '@testing-library/react';
import Orders from './Orders';
import GlobalContexts from '../../../contexts/globalContexts';
import { mockProducts } from '../../../constants/mocks';

describe('Test <Orders /> component', () => {
  let contexts;
  let renderOrders;

  beforeEach(() => {
    contexts = {
      products: mockProducts,
    };

    renderOrders = render(
      <GlobalContexts.Provider value={{contexts}}>
        <Orders />
      </GlobalContexts.Provider>
    );
  });

  afterEach(cleanup);

  it('Orders Snapshot test', () => {
    const { asFragment } = renderOrders;

    expect(asFragment(renderOrders)).toMatchSnapshot();
  });

  it('should disable "quantity minus button" when quantity is 0', async () => {
    // Act
    const minusButtonText = await screen.findAllByText('-');

    // Assert
    expect(minusButtonText[0].className.includes("disabled")).toBeTruthy();
    expect(minusButtonText.length > 0).toBeTruthy();
  });

  it('should update the total value after click "quantity plus button"', async () => {
    const onePlusButton = await screen.findAllByText('+');

    fireEvent(onePlusButton[0],
      new MouseEvent('click', {
        bubbles: true,
        cancelable: true,
      }),
    );

    const totalCost = screen.getByTestId('order-total').innerHTML.split('$')[1];

    expect(parseFloat(totalCost) > 0).toBeTruthy();
  });
});
