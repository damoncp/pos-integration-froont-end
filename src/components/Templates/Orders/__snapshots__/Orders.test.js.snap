// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test <Orders /> component Orders Snapshot test 1`] = `
<DocumentFragment>
  <div
    id="orders"
  >
    <h5
      class="text--centered text--bold heading__default"
    >
      Order
    </h5>
    <div
      class="container"
    >
      <div
        class="row"
      >
        <div
          class="col"
        >
          <div
            class="table-responsive"
          >
            <table
              class="table"
            >
              <thead>
                <tr>
                  <th>
                    Quantity
                  </th>
                  <th>
                    Name
                  </th>
                  <th>
                    Price
                  </th>
                </tr>
              </thead>
              <tbody>
                <tr
                  class="orders-table__row"
                >
                  <td>
                    <button
                      class="btn btn-secondary disabled"
                      disabled=""
                      type="button"
                    >
                      -
                    </button>
                    <span
                      class="orders-text__regular"
                    >
                      0
                    </span>
                    <button
                      class="btn btn-secondary"
                      type="button"
                    >
                      +
                    </button>
                  </td>
                  <td
                    class="orders-text__regular"
                  >
                    Chicken Burger
                  </td>
                  <td
                    class="orders-text__regular"
                  >
                    $12
                  </td>
                </tr>
                <tr
                  class="orders-table__row"
                >
                  <td>
                    <button
                      class="btn btn-secondary disabled"
                      disabled=""
                      type="button"
                    >
                      -
                    </button>
                    <span
                      class="orders-text__regular"
                    >
                      0
                    </span>
                    <button
                      class="btn btn-secondary"
                      type="button"
                    >
                      +
                    </button>
                  </td>
                  <td
                    class="orders-text__regular"
                  >
                    Vegan Burger
                  </td>
                  <td
                    class="orders-text__regular"
                  >
                    $12
                  </td>
                </tr>
                <tr
                  class="orders-table__row"
                >
                  <td>
                    <button
                      class="btn btn-secondary disabled"
                      disabled=""
                      type="button"
                    >
                      -
                    </button>
                    <span
                      class="orders-text__regular"
                    >
                      0
                    </span>
                    <button
                      class="btn btn-secondary"
                      type="button"
                    >
                      +
                    </button>
                  </td>
                  <td
                    class="orders-text__regular"
                  >
                    Beef Burger
                  </td>
                  <td
                    class="orders-text__regular"
                  >
                    $12
                  </td>
                </tr>
                <tr
                  class="orders-table__row"
                >
                  <td>
                    <button
                      class="btn btn-secondary disabled"
                      disabled=""
                      type="button"
                    >
                      -
                    </button>
                    <span
                      class="orders-text__regular"
                    >
                      0
                    </span>
                    <button
                      class="btn btn-secondary"
                      type="button"
                    >
                      +
                    </button>
                  </td>
                  <td
                    class="orders-text__regular"
                  >
                    Lamb Burger
                  </td>
                  <td
                    class="orders-text__regular"
                  >
                    $12
                  </td>
                </tr>
                <tr
                  class="orders-table__row"
                >
                  <td>
                    <button
                      class="btn btn-secondary disabled"
                      disabled=""
                      type="button"
                    >
                      -
                    </button>
                    <span
                      class="orders-text__regular"
                    >
                      0
                    </span>
                    <button
                      class="btn btn-secondary"
                      type="button"
                    >
                      +
                    </button>
                  </td>
                  <td
                    class="orders-text__regular"
                  >
                    Small Fries
                  </td>
                  <td
                    class="orders-text__regular"
                  >
                    $7
                  </td>
                </tr>
                <tr
                  class="orders-table__row"
                >
                  <td>
                    <button
                      class="btn btn-secondary disabled"
                      disabled=""
                      type="button"
                    >
                      -
                    </button>
                    <span
                      class="orders-text__regular"
                    >
                      0
                    </span>
                    <button
                      class="btn btn-secondary"
                      type="button"
                    >
                      +
                    </button>
                  </td>
                  <td
                    class="orders-text__regular"
                  >
                    Large Fries
                  </td>
                  <td
                    class="orders-text__regular"
                  >
                    $7
                  </td>
                </tr>
                <tr
                  class="orders-table__row"
                >
                  <td>
                    <button
                      class="btn btn-secondary disabled"
                      disabled=""
                      type="button"
                    >
                      -
                    </button>
                    <span
                      class="orders-text__regular"
                    >
                      0
                    </span>
                    <button
                      class="btn btn-secondary"
                      type="button"
                    >
                      +
                    </button>
                  </td>
                  <td
                    class="orders-text__regular"
                  >
                    Coca Cola
                  </td>
                  <td
                    class="orders-text__regular"
                  >
                    $3
                  </td>
                </tr>
              </tbody>
            </table>
          </div>
        </div>
      </div>
      <div
        class="row"
      >
        <div
          class="col"
        >
          Total: 
        </div>
        <div
          class="col"
          data-testid="order-total"
        >
          $ 0
        </div>
      </div>
      <div
        class="row"
      >
        <button
          class="element--no-border-radius bg-success text-white p-3 mt-2 btn btn-secondary disabled"
          disabled=""
          type="button"
        >
          Pay
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
